### PR TITLE
Fix imports for images with no info about MediaType.

### PR DESCRIPTION
**Pull request by tomob**
Signed-off-by: Tomasz Baranski <tbaransk@redhat.com>

**What this PR does / why we need it**:

An image from a registry might have MediaType for its layer empty. This
PR adds code which tries to guess the MediaType basing on so-called
magic numbers aka file signatures.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/image/transport.go [hunk 1]
```diff
@@ -129,6 +133,71 @@ func isDir(path string) bool {
...
+		head[0x105] == 0x72
+}
+
+func isFiletypeGzip(head []byte) bool {
+	// Signature from https://en.wikipedia.org/wiki/List_of_file_signatures
+	return len(head) > 1 &&
+		head[0] == 0x1F && head[1] == 0x8B
+}
+
+func guessMediaType(ctx context.Context, src types.ImageSource, layer types.BlobInfo, cache types.BlobInfoCache) (string, error) {
```
Comment (awels) [inline pkg/image/transport.go:152]: This feels like it is doing the same as https://github.com/kubevirt/containerized-data-importer/blob/master/pkg/importer/format-readers.go#L227 more or less. But I can't come up with a good way of using that here of the top of my head.

Comment (tomob) [reply pkg/image/transport.go:152]: Oh, I didn't know about this code.

I've given it some thought. Maybe we could get rid of the whole `MediaType` checking and just use FormatReaders? If a layer can be read using FR, it's good enough for the import process. The declared edia type wouldn't matter.

What do you think?

Comment (awels) [reply pkg/image/transport.go:152]: Yeah I think that would be good. I just hate re-implementing code we have been using successfully for some time now. Also maybe some unit tests to verify it is working as expected?

**Review by awels — APPROVED** [review]
/lgtm
/approve

**Comment by awels** [comment]
looks like there are some failing tests now....

**Comment by tomob** [comment]
> looks like there are some failing tests now....

This is weird... With `FormatReaders` reading is very sloooooow. But I can't see any reason for it.

File: pkg/importer/transport.go [hunk 2]
```diff
@@ -36,15 +34,11 @@ import (
...
-	dockerTarLayerMediaType = "application/vnd.docker.image.rootfs.diff.tar"
-	ociTarLayerMediaType    = "application/vnd.oci.image.layer.v1.tar"
-	whFilePrefix            = ".wh."
+	whFilePrefix = ".wh."
 )
 
 func commandTimeoutContext() (context.Context, context.CancelFunc) {
-	ctx := context.Background()
-	return context.WithTimeout(ctx, cancelTimeout)
+	return context.WithCancel(context.Background())
```
Comment (tomob) [inline pkg/importer/transport.go:41]: @awels I think we cannot say what a "generic" timeout for any situation should be. So I got rid of it; the importer will now patiently download images for as long as it takes.

File: tools/cdi-func-test-file-host-init/nginx.conf [hunk 3]
```diff
@@ -120,7 +120,7 @@ http {
             proxy_set_header X-Forwarded-Proto "https";
             proxy_read_timeout                 900;
 
-            limit_rate 100k;
+            limit_rate 1M;
```
Comment (tomob) [inline tools/cdi-func-test-file-host-init/nginx.conf:123]: @awels It seems 100k (per second) was too harsh. The test was taking ages to run.

**Review by awels — COMMENTED** [review]
One question, you mentioned with the FormatReaders it was much slower. How did we fix that? Hopefully not by just waiting longer. Or was the timeout causing a restart and that is why it was slower?

**Comment by tomob** [comment]
> One question, you mentioned with the FormatReaders it was much slower. How did we fix that? Hopefully not by just waiting longer. Or was the timeout causing a restart and that is why it was slower?

No, I found the bug in the old code (the one I replaced) which made it skip some image layers (their media type was `...tar`, while it actually was gzipped, so the type should have been `....tar+gz`). After I fixed that both versions took about the same.

So, it's not that FormatReaders are slower – it's that my code that was too fast because it skipped layers it should process.

**Comment by awels** [comment]
Okay so by switching to code we knew was working we fixed a bug before it became a bug. Nice

**Pull request merged**